### DISCUSSION TITLE
Problem with take_sample() Function

### DISCUSSION THREAD

**Discussion by xdodgerx** [zubkov-andrei/pg_profile#46] [Q&A]
Добрый день. Установлен релиз 0.3.6 (последний).

После некоторого времени работы на сервере при выполнении take_sample стала появляться ошибка:
(local,"повторяющееся значение ключа нарушает ограничение уникальности ""pk_last_stat_tables""           +
 SQL-оператор: ""INSERT INTO last_stat_tables.....)
 )""                                                                                                +
 функция PL/pgSQL collect_obj_stats(jsonb,integer,integer,text,boolean), строка 179, оператор SQL-оператор+
 функция PL/pgSQL take_sample(integer,boolean), строка 882, оператор присваивание                         +
 функция PL/pgSQL take_sample_subset(integer,integer), строка 27, оператор присваивание                   +
 SQL-функция ""take_sample"", оператор 1                                                                  +
 Ключ ""(server_id, sample_id, datid, relid)=(1, 1, 931582, 939957)"" уже существует.",00:00:00.45)

Сама таблица last_stat_tables, разумеется, пустая. 

При этом несколько раз переустанавливал pg_profile, ставил в разные схемы и т.д.

Не подскажете, в чём может быть проблема?

PostgreSQL 13.6 (Ubuntu 13.6-1.pgdg20.04+1) on x86_64-pc-linux-gnu, compiled by gcc (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0, 64-bit

**Comment by zubkov-andrei** [comment]
Добрый день!
Очень похоже на #28. Проверьте состояние pg_statio_all_tables для этого отношения. И если там есть дубликаты, ищите инвалидные индексы на таблице. После их удаления проблема должна решиться.

**Reply by xdodgerx** [reply]
Спасибо большое, это именно оно. Оставлю тут краткое руководство, с Вашего разрешения!

Сначала найдём объект, у которого появилась 2я запись: 
select COUNT (*), schemaname, relname from pg_statio_all_tables group by schemaname, relname order by 1 DESC\gx
Данный запрос выдаст нужные объекты (у них COUNT>1)

Далее выполняем для этого объекта:
REINDEX TABLE "имя объекта"

Получаем сообщение об ошибке типа: 
перестроить нерабочий индекс "pg_toast.pg_toast_939957_index_ccnew1" в таблице TOAST нельзя, он пропускается

Выполняем: DROP INDEX pg_toast.pg_toast_939957_index_ccnew1

И снова: REINDEX TABLE "имя объекта"

После этого всё заработало.
